engine/engine/content/builtins/scripts/edn.lua

```diff
@@ -111,57 +94,60 @@ local function encode_as_primitive(val)
   return encoder(val)
 end
 
-local function collect_refs(depth, val, refs)
-  local t = type(val)
-  if t == "table" then
-    if depth < 100 and not refs[val] then
-      refs[val] = val
-      local next_depth = depth + 1
-      for k, v in pairs(val) do
-        collect_refs(next_depth, k, refs)
-        collect_refs(next_depth, v, refs)
-      end
-    end
-  end
-  return refs
-end
-
 local function encode_table(val)
   local encoded_kvs = {}
   for k, v in pairs(val) do
-    table.insert(
-      encoded_kvs,
-      encode_as_primitive(k) .. " " .. encode_as_primitive(v))
+    encoded_kvs[#encoded_kvs + 1] = encode_as_primitive(k) .. " " .. encode_as_primitive(v)
   end
   return "#lua/table{:content [" .. table.concat(encoded_kvs, " ") .. "] :string " .. encode_string(str(val)) .. "}"
 end
 
 local function encode_refs(refs)
   local encoded_refs = {}
-  for k, v in pairs(refs) do
-    table.insert(
-      encoded_refs,
-      encode_as_primitive(k) .. " " .. encode_table(v))
+  local val
+  for i = 1, #refs do
+    val = refs[i]
+    encoded_refs[i] = encode_as_primitive(val).." "..encode_table(val)
```

Comment:
Here we are talking about 450 000 - 500 000 values, on such numbers `fori` MUCH faster then `forp`
To be able to use it I added a separate table `dups` for duplicates for `collect_refs`

Reply:
I don't understand why we are calling both encode_as_primitive() and encode_table() on `val`?

The old code did `encode_as_primitive(k)` on the table key. What is the key with this change?

Reply:
yes, but old table was
refs[val] = val
so the key was equal to the value

Reply:
Oh, ok, weird? But then I guess it is fine.